pkg/cmd/cli/cmd/debug.go

```diff
@@ -0,0 +1,565 @@
+				if s.Name != containerName {
+					continue
+				}
+				return s.State.Running != nil, nil
+			}
+			return false, nil
+		}
+		return false, nil
+	}
+}
+
+// SingleObject returns a ListOptions for watching a single object.
+// TODO: move to pkg/api/helpers.go upstream.
+func SingleObject(meta kapi.ObjectMeta) kapi.ListOptions {
+	return kapi.ListOptions{
+		FieldSelector:   fields.OneTermEqualSelector("metadata.name", meta.Name),
+		ResourceVersion: meta.ResourceVersion,
+	}
+}
+
+// Debug creates and runs a debugging pod.
+func (o *DebugOptions) Debug() error {
+	pod, originalCommand := o.transformPodForDebug(o.Annotations)
+	var commandString string
+	switch {
+	case len(originalCommand) > 0:
+		commandString = strings.Join(originalCommand, " ")
+	default:
+		commandString = "<image entrypoint>"
+	}
+
+	if o.Print != nil {
+		return o.Print(pod, o.Attach.Out)
+	}
+
+	glog.V(5).Infof("Creating pod: %#v", pod)
+	fmt.Fprintf(o.Attach.Err, "Debugging with pod/%s, original command: %s\n", pod.Name, commandString)
+	pod, err := o.createPod(pod)
+	if err != nil {
+		return err
+	}
+
+	// ensure the pod is cleaned up on shutdown
+	o.Attach.InterruptParent = interrupt.New(
+		func(os.Signal) { os.Exit(1) },
+		func() {
+			fmt.Fprintf(o.Attach.Out, "\nRemoving debug pod ...\n")
+			if err := o.Attach.Client.Pods(pod.Namespace).Delete(pod.Name, kapi.NewDeleteOptions(0)); err != nil {
+				fmt.Fprintf(o.Attach.Err, "error: unable to delete the debug pod %q: %v", pod.Name, err)
+			}
+		},
+	)
+	glog.V(5).Infof("Created attach arguments: %#v", o.Attach)
+	return o.Attach.InterruptParent.Run(func() error {
+		w, err := o.Attach.Client.Pods(pod.Namespace).Watch(SingleObject(pod.ObjectMeta))
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(o.Attach.Err, "Waiting for pod to start ...\n")
+		switch _, err := Until(o.Timeout, w, PodContainerRunning(o.Attach.ContainerName)); {
+		// switch to logging output
+		case err == ErrPodCompleted, !o.Attach.Stdin:
+			_, err := kcmd.LogsOptions{
+				Object: pod,
+				Options: &kapi.PodLogOptions{
+					Container: o.Attach.ContainerName,
+					Follow:    true,
+				},
+				Out: o.Attach.Out,
+
+				LogsForObject: o.LogsForObject,
+			}.RunLogs()
+			return err
+		case err != nil:
+			return err
+		default:
+			// TODO: attach can race with pod completion, allow attach to switch to logs
+			return o.Attach.Run()
+		}
+	})
+}
+
+// transformPodForDebug alters the input pod to be debuggable
+func (o *DebugOptions) transformPodForDebug(annotations map[string]string) (*kapi.Pod, []string) {
+	pod := o.Attach.Pod
+
+	// reset the container
+	container := containerForName(pod, o.Attach.ContainerName)
+
+	// identify the command to be run
+	originalCommand := append(container.Command, container.Args...)
+	container.Command = o.Command
+	if len(originalCommand) == 0 {
+		if cmd, ok := imageapi.ContainerImageEntrypointByAnnotation(pod.Annotations, o.Attach.ContainerName); ok {
+			originalCommand = cmd
```

Comment:
For the command, we print it so you can cut and paste.  I think that is as
close as we can get safely.

Removing keep labels for now.  Will follow up with port-forward post 3.2?

On Fri, Feb 19, 2016 at 4:56 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I do as well, but I don't have a safe, predictable way to quote down to
> bash safely.  Punt and come back?
>
> On Fri, Feb 19, 2016 at 4:16 PM, David Eads <notifications@github.com>
> wrote:
>
>> In pkg/cmd/cli/cmd/debug.go
>> <https://github.com/openshift/origin/pull/7294#discussion_r53519300>:
>>
>> > +	})
>> > +}
>> > +
>> > +// transformPodForDebug alters the input pod to be debuggable
>> > +func (o *DebugOptions) transformPodForDebug(annotations map[string]string) (*kapi.Pod, []string) {
>> > +	pod := o.Attach.Pod
>> > +
>> > +	// reset the container
>> > +	container := containerForName(pod, o.Attach.ContainerName)
>> > +
>> > +	// identify the command to be run
>> > +	originalCommand := append(container.Command, container.Args...)
>> > +	container.Command = o.Command
>> > +	if len(originalCommand) == 0 {
>> > +		if cmd, ok := imageapi.ContainerImageEntrypointByAnnotation(pod.Annotations, o.Attach.ContainerName); ok {
>> > +			originalCommand = cmd
>>
>> I think I'd like this as an env var in my container so that I can just
>> execute it to try it.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/7294/files#r53519300>.
>>
>
>